src/main/java/com/vmware/vipclient/i18n/messages/service/ProductService.java

```diff
@@ -44,23 +56,63 @@ public JSONArray getSupportedLocalesFromRemoteVIP() {
 
     public List<Map> getAllComponentTranslation() {
         List<Map> list = new ArrayList<Map>();
-        Object[] locales = {};
-        Object[] components = {};
-        if (VIPCfg.getInstance().getMessageOrigin() == DataSourceEnum.VIP) {
-            locales = this.getSupportedLocalesFromRemoteVIP().toArray();
-            components = this.getComponentsFromRemoteVIP()
-                    .toArray();
-        }
-        for (Object locale : locales) {
-            for (Object component : components) {
-                dto.setComponent(((String) component).trim());
-                dto.setLocale(LocaleUtility.fmtToMappedLocale((String) locale).toString().trim());
-                Map<String, String> retMap = new ComponentService(dto).getMessages().getCachedData();
-                if (retMap != null) {
-                    list.add(retMap);
+        Map<String, String> locales = this.getLanguages(VIPCfg.getInstance().getMsgOriginsQueue().iterator());
+        List<String> components = this.getComponents(VIPCfg.getInstance().getMsgOriginsQueue().iterator());
+        if (locales != null) {
+            for (String languageTag : locales.keySet()) {
+                for (Object component : components) {
+                    dto.setComponent(((String) component).trim());
+                    dto.setLocale(LocaleUtility.fmtToMappedLocale(Locale.forLanguageTag(languageTag)).toString().trim());
+                    dto.setProductID(VIPCfg.getInstance().getProductName());
+                    dto.setVersion(VIPCfg.getInstance().getVersion());
+                    Map<String, String> retMap = new ComponentService(dto).getMessages().getCachedData();
+                    if (retMap != null) {
+                        list.add(retMap);
+                    }
                 }
             }
         }
         return list;
     }
+
+    private Map<String, String> getLanguages(Iterator<DataSourceEnum> msgSourceQueueIter) {
+        if (!msgSourceQueueIter.hasNext())
+            return null;
+
+        DataSourceEnum dataSource = msgSourceQueueIter.next();
+        LocaleOpt opt = dataSource.createLocaleOpt();
+        Map<String, String> languages =  opt.getLanguages(LocaleUtility.getDefaultLocale().toLanguageTag());
```

Comment:
@Xiaochao8 , what do you mean by "get English locale names from bundles"? We do not even have the English display name inside the bundle. 

I use Java's Locale.forLanguageTag to get the display name in case of offline mode: `Locale inLocale = Locale.forLanguageTag(displayLanguage); `

Reply:
We get supported locales from bundle file names trimmed prefix 'messages_' and suffix '.json'. This has nothing to do with display language. It's pure English and must be consistent with bundle file names.
These locale names are used to query translation.

Reply:
In a bundle named messages_en.json, the suffix 'en' is the "language tag", not the "display name". The display name would be "English" if the **display language** is "en". If the display language is "es", the display name would be "Ingles".
So where in the bundle file can we get the word "English"?
And where in the bundle file can we get "Ingles" if the display language is es?

What is "pure English" that you are talking about? The bundle file name only has language tags, not display names. The language tag "en" can be displayed as "English", "Ingles", "Anglais", etc..

Reply:
To further explain "dusplay names" top you, here is the getSupportedLocales response from the service:

`"languages": [
      {
        "languageTag": "en",
        "displayName": "inglés",
        "displayName_sentenceBeginning": "Inglés",
        "displayName_uiListOrMenu": "Inglés",
        "displayName_standalone": "Inglés"
      },
      {
        "languageTag": "de",
        "displayName": "alemán",
        "displayName_sentenceBeginning": "Alemán",
        "displayName_uiListOrMenu": "Alemán",
        "displayName_standalone": "Alemán"
      },
      {
        "languageTag": "zh-Hans",
        "displayName": "chino simplificado",
        "displayName_sentenceBeginning": "Chino simplificado",
        "displayName_uiListOrMenu": "Chino simplificado",
        "displayName_standalone": "Chino simplificado"
      },
      {
        "languageTag": "ja",
        "displayName": "japonés",
        "displayName_sentenceBeginning": "Japonés",
        "displayName_uiListOrMenu": "Japonés",
        "displayName_standalone": "Japonés"
      },
      {
        "languageTag": "zh-Hant",
        "displayName": "chino tradicional",
        "displayName_sentenceBeginning": "Chino tradicional",
        "displayName_uiListOrMenu": "Chino tradicional",
        "displayName_standalone": "Chino tradicional"
      },
      {
        "languageTag": "es",
        "displayName": "español",
        "displayName_sentenceBeginning": "Español",
        "displayName_uiListOrMenu": "Español",
        "displayName_standalone": "Español"
      },
      {
        "languageTag": "ko",
        "displayName": "coreano",
        "displayName_sentenceBeginning": "Coreano",
        "displayName_uiListOrMenu": "Coreano",
        "displayName_standalone": "Coreano"
      },
      {
        "languageTag": "fr",
        "displayName": "francés",
        "displayName_sentenceBeginning": "Francés",
        "displayName_uiListOrMenu": "Francés",
        "displayName_standalone": "Francés"
      },
      {
        "languageTag": "fr-CA",
        "displayName": "francés canadiense",
        "displayName_sentenceBeginning": "Francés canadiense",
        "displayName_uiListOrMenu": "Francés canadiense",
        "displayName_standalone": "Francés canadiense"
      }
    ],`

In a similar fashion, the offline mode code also gives a display name, only it is simplified by what we get using `Locale inLocale = Locale.forLanguageTag(displayLanguage)`

Reply:
I think you are totally lost. In this interface, the only needed is the locale names from bundle file names. That's what I said: bundle file names trim prefix 'messages_' and suffix '.json'. This has nothing to do with display language.  Don't mix different things together to make things become complex. Forget display language here.